File: algorithms/First Bad Version/First Bad Version.py
class Solution:
    def firstBadVersion(self, n: int) -> int:
        fast, slow = int(n/2), n
        diff = abs(fast-slow)
        while isBadVersion(fast) == isBadVersion(slow) or diff > 1:
            fast, slow = fast + (-1)**isBadVersion(fast) * (int(diff/2) or 1), fast
            diff = abs(fast-slow)
        return fast if isBadVersion(fast) else slow

File: scripts/algorithms/F/First Bad Version/First Bad Version.js
// Runtime: 109 ms (Top 17.20%) | Memory: 41.9 MB (Top 54.94%)
/**
 * Definition for isBadVersion()
 *
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 * ...
 * };
 */

/**
 * @param {function} isBadVersion()
 * @return {function}
 */
var solution = function(isBadVersion) {
    /**
     * @param {integer} n Total versions
     * @return {integer} The first bad version
     */
    return function(n) {
        let ceiling = n
        let floor = 1
        let firstBadVersion = -1

        while (floor <= ceiling) {
            const middle = Math.floor((ceiling + floor) / 2)

            if (isBadVersion(middle)) {

                firstBadVersion = middle
                ceiling = middle - 1
            } else {
                floor = middle + 1
            }
        }

        return firstBadVersion
    };
};